File: textbook/python/01-dataflow/01_dataflow_model.py
# Copyright 2026 Terrene Foundation
# SPDX-License-Identifier: Apache-2.0
"""
# ════════════════════════════════════════════════════════════════════════
# Kailash SDK Textbook — DataFlow / Model Definition
# ════════════════════════════════════════════════════════════════════════
# OBJECTIVE: Define database models using @db.model decorator and field types
# LEVEL: Basic
# PARITY: Full — Rust uses ModelDefinition::new() with FieldType enum;
#         Python uses @db.model decorator with type annotations
# VALIDATES: DataFlow(), @db.model decorator, field type mapping, initialize()
#
# Run: uv run python textbook/python/01-dataflow/01_dataflow_model.py
# ════════════════════════════════════════════════════════════════════════
"""
from __future__ import annotations

import asyncio
import os
import tempfile

from dataflow import DataFlow

# ── 1. Create a DataFlow instance ───────────────────────────────────
# DataFlow connects to a database. SQLite is simplest for tutorials.
# In production, use PostgreSQL via DATABASE_URL environment variable.

db_path = os.path.join(tempfile.gettempdir(), "textbook_dataflow_01.db")
db = DataFlow(database_url=f"sqlite:///{db_path}")

# ── 2. Define a model with @db.model ───────────────────────────────
# The decorator registers the class, creates CRUD nodes, and maps
# Python type annotations to database column types.
#
# Supported types: str, int, float, bool, datetime
# Fields with defaults are optional; fields without are required.


@db.model
class Product:
    name: str
    price: float
    category: str
    in_stock: bool = True


# Verify the model was registered
registered_models = db.get_models()
assert "Product" in registered_models, "Product model should be registered"

# Inspect model fields
fields = db.get_model_fields("Product")
assert "name" in fields, "name field registered"
assert "price" in fields, "price field registered"
assert "category" in fields, "category field registered"
assert "in_stock" in fields, "in_stock field registered"

# Required vs optional: fields with defaults are not required
assert fields["name"]["required"] is True, "name has no default — required"
assert fields["in_stock"]["required"] is False, "in_stock has default — optional"
assert fields["in_stock"]["default"] is True

print(f"Registered models: {list(registered_models.keys())}")
print(f"Product fields: {list(fields.keys())}")

# ── 3. Multiple models ─────────────────────────────────────────────


@db.model
class Customer:
    email: str
    name: str
    tier: str = "free"


@db.model
class Order:
    customer_email: str
    product_name: str
    quantity: int
    total: float


all_models = db.get_models()
assert len(all_models) >= 3, "Three models registered"
assert "Customer" in all_models
assert "Order" in all_models

# ── 4. Initialize — create tables in the database ──────────────────
# initialize() is async — it creates tables, runs migrations,
# and sets up the connection pool.


async def main():
    success = await db.initialize()
    assert success, "DataFlow initialization should succeed"

    # After initialize, tables exist and CRUD is available
    print(f"DataFlow initialized with {len(all_models)} models")

    # ── 5. Table name mapping ───────────────────────────────────────
    # By default, class names are converted to snake_case table names:
    #   Product → products
    #   Customer → customers
    #   Order → orders
    # Override with __tablename__ class attribute if needed.

    # Cleanup
    await db.stop()


asyncio.run(main())

# ── 6. Edge case: custom table name ────────────────────────────────

db2 = DataFlow(database_url=f"sqlite:///{db_path}")


@db2.model
class AuditLog:
    __tablename__ = "custom_audit_logs"
    action: str
    details: str


# The table name override is respected
model_info = db2._models.get("AuditLog", {})
assert model_info.get("table_name") == "custom_audit_logs"

# Cleanup temp file
try:
    os.unlink(db_path)
except OSError:
    pass

print("PASS: 01-dataflow/01_dataflow_model")

File: textbook/rust/01-dataflow/src/bin/01_dataflow_model.rs
// Copyright 2026 Terrene Foundation
// SPDX-License-Identifier: Apache-2.0
//! Kailash SDK Textbook -- DataFlow / Model Definition
//!
//! OBJECTIVE: Define database models using ModelDefinition builder and FieldType enum.
//! LEVEL: Basic
//! PARITY: Full -- Python uses @db.model decorator with type annotations;
//!         Rust uses ModelDefinition::new() with builder pattern and FieldType enum.
//! VALIDATES: DataFlow, ModelDefinition, FieldType, FieldDef, register_model
//!
//! Run: cargo run -p tutorial-dataflow --bin 01_dataflow_model

use kailash_dataflow::prelude::*;

#[tokio::main]
async fn main() {
    // ── 1. Create a DataFlow Instance ──
    // DataFlow connects to a database. SQLite in-memory is simplest for
    // tutorials. In production, use PostgreSQL via a connection string.

    let mut df = DataFlow::new("sqlite::memory:")
        .await
        .expect("connect to sqlite");

    // ── 2. Define a Model ──
    // ModelDefinition describes a table: name, table name, and fields.
    // Each field has a name, type, and optional configuration via closure.
    //
    // Python equivalent:
    //   @db.model
    //   class User:
    //       id: int = field(primary_key=True)
    //       name: str
    //       email: str

    let user_model = ModelDefinition::new("User", "users")
        .field("id", FieldType::Integer, |f| f.primary_key())
        .field("name", FieldType::Text, |f| f.required())
        .field("email", FieldType::Text, |f| f.required().unique())
        .field("age", FieldType::Integer, |f| f)
        .auto_timestamps();

    // auto_timestamps() adds created_at and updated_at fields automatically.
    // These are managed by the framework -- never set them manually.

    // ── 3. Inspect Model Metadata ──
    // The model carries metadata about its fields before registration.

    assert_eq!(user_model.name(), "User");
    assert_eq!(user_model.table_name(), "users");

    let fields = user_model.fields();
    assert!(fields.len() >= 4); // id, name, email, age (+ timestamps)

    // ── 4. FieldType Enum ──
    // FieldType maps to SQL column types via the dialect layer.
    //
    //   Integer  -> INTEGER / INT
    //   Text     -> TEXT / VARCHAR
    //   Boolean  -> BOOLEAN / INT (SQLite)
    //   Float    -> REAL / DOUBLE
    //   DateTime -> DATETIME / TIMESTAMP
    //   Json     -> JSON / TEXT (SQLite)
    //   Blob     -> BLOB / BYTEA

    let product_model = ModelDefinition::new("Product", "products")
        .field("id", FieldType::Integer, |f| f.primary_key())
        .field("name", FieldType::Text, |f| f.required())
        .field("price", FieldType::Float, |f| f.required())
        .field("active", FieldType::Boolean, |f| f.default_value("true"))
        .field("metadata", FieldType::Json, |f| f)
        .field("image", FieldType::Blob, |f| f)
        .auto_timestamps();

    assert_eq!(product_model.name(), "Product");

    // ── 5. Register Models ──
    // register_model() validates the definition and creates the table.
    // After registration, 11 CRUD/bulk workflow nodes are available:
    //   CreateUser, ReadUser, UpdateUser, DeleteUser, ListUser,
    //   UpsertUser, CountUser, BulkCreateUser, BulkUpdateUser,
    //   BulkDeleteUser, BulkUpsertUser

    df.register_model(user_model).expect("register User");
    df.register_model(product_model).expect("register Product");

    // ── 6. Model with Foreign Key ──
    // ForeignKeyRef links a field to another model's field.

    let order_model = ModelDefinition::new("Order", "orders")
        .field("id", FieldType::Integer, |f| f.primary_key())
        .field("user_id", FieldType::Integer, |f| {
            f.required().foreign_key(ForeignKeyRef::new("users", "id"))
        })
        .field("total", FieldType::Float, |f| f.required())
        .field("status", FieldType::Text, |f| {
            f.required().default_value("pending")
        })
        .auto_timestamps();

    df.register_model(order_model).expect("register Order");

    // ── 7. Verify Registration ──
    // After registration, models are accessible via the DataFlow instance.

    let registered = df.registered_models();
    assert!(registered.contains(&"User".to_string()));
    assert!(registered.contains(&"Product".to_string()));
    assert!(registered.contains(&"Order".to_string()));
    assert_eq!(registered.len(), 3);

    println!("PASS: 01-dataflow/01_dataflow_model");
}
